Fix Debian 8 support: configure for systemd

JayH5 opened the pull request:
Debian 8 is listed as a supported platform in `metadata.json`, but the logic to detect systemd configuration seems to expect only Ubuntu within the Debian family. This checks for vanilla Debian vs Ubuntu and applies systemd configuration as necessary.

<!--
Thank you for contributing to this project!

- This project has a Contributor Code of Conduct: https://voxpupuli.org/coc/
- Please check that here is no existing issue or PR that addresses your problem.
- Our vulnerabilities reporting process is at https://voxpupuli.org/security/

-->

JayH5 commented:
I'm not sure what's going on with those Arch Linux test failures...that isn't a part of the code that I changed.

juniorsysadmin commented on manifests/config.pp:
```diff
@@ -193,28 +193,22 @@
 
   case $facts['os']['family'] {
+    'Archlinux': {
+      $has_systemd = true
```
Can we perhaps make use of the systemd fact in camptopcamp/systemd module?

JayH5 replied:
I don't know much about that fact.. if it works accurately it would be cool to use. I can add the module as a dependency? It would also be useful for the systemctl daemon reload stuff.

bastelfreak replied:
yes, you can add it as a dependency to the metadata.json and use the fact

wyardley commented:
FWIW, I'm seeing those Arch test failures on current master as well. Paging Archlinux evangelist @bastelfreak to the white courtesy phone.

bastelfreak commented:
yes this is related to the new facts

bastelfreak commented:
@JayH5 I will have a look at the spec failures related to archlinux and fix them. I maybe already need to add the dependency to the systemd module. You can hold on with further development here until I fixed the current issues.

bastelfreak commented:
@JayH5 can you rebase please? I fixed spec errors in https://github.com/voxpupuli/puppet-rabbitmq/pull/685

JayH5 commented:
Did a quick rebase but there are errors now... I'm a little swamped with other things at the moment and might only get back to this next week.

JayH5 commented:
I've tried to make use of the systemd module. I hit a few bumps and it has taken me a while to get this working. I've run out of time to spend on this for and the tests still aren't passing 😢.

I've done the bulk of the work here and would be grateful if somebody could take over. Otherwise, I will only be able to revisit again sometime in the future (1 week to 1 month's time).

wyardley commented:
`bundle exec rake rubocop:auto_correct` will fix the linter errors.
I have played a little with trying to use systemd module to manage limits in the past, and have not had the greatest of luck (though I thought this particular problem had already been resolved), because the types are a bit too restrictive. This is one reason why this module is still not using the camptocamp systemd module.

Currently, LimitNOFILE can still only be integer:
https://github.com/camptocamp/puppet-systemd/blob/master/types/servicelimits.pp#L10

I'll see if I can submit a PR to fix this; until then, I don't think we can use this module to manage the systemd service limits. Even then, it would not be ideal, because folks would have to use the newest version.

```
      Puppet::PreformattedError:
        Evaluation Error: Error while evaluating a Resource Statement, Systemd::Service_limits[rabbitmq.service]: parameter 'limits' entry 'LimitNOFILE' expects an Integer[0] value, got Integer[-1, -1] (file: /Users/xx/git/puppet-rabbitmq/spec/fixtures/modules/rabbitmq/manifests/config.pp, line: 220) on node xxx
```

wyardley commented:
https://github.com/camptocamp/puppet-systemd/pull/76

bastelfreak commented on manifests/config.pp:
```diff
@@ -254,28 +212,14 @@
+    default: { }
+  }
+
+  if $::facts['systemd'] { # systemd fact provided by systemd module
```
We should not use topscope anymore. Please use `$facts['systemd']`

bastelfreak commented on spec/classes/rabbitmq_spec.rb:
```diff
@@ -1425,6 +1413,13 @@
         }
       end
 
+      context 'on systems with systemd', if: facts[:systemd] do
```
nice, I always forget that we can set conditions within a context block

bastelfreak commented on manifests/service.pp:
```diff
@@ -32,6 +32,10 @@
       name       => $service_name,
     }
+
+    if $::facts['systemd'] {
```
please use `$facts['systemd']`

JayH5 commented:
Rebased and updated as suggested. I'm still stumped by the Arch Linux test failures...if anybody has any ideas...

Once again, don't have a heck of a lot of time to work on this so may be slow to update and happy for others to take over.

bastelfreak commented:
@JayH5 the errors are not archlinux specific:

```
rspec ./spec/classes/rabbitmq_spec.rb[1:2:10:1] # rabbitmq on archlinux-4-x86_64 with file_limit => 'unlimited' should not contain File[/etc/security/limits.d/rabbitmq-server.conf]
rspec ./spec/classes/rabbitmq_spec.rb[1:2:10:2] # rabbitmq on archlinux-4-x86_64 with file_limit => 'unlimited' should not contain File[/etc/default/rabbitmq-server]
rspec ./spec/classes/rabbitmq_spec.rb[1:2:10:3] # rabbitmq on archlinux-4-x86_64 with file_limit => 'unlimited' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:2:11:1] # rabbitmq on archlinux-4-x86_64 with file_limit => 'infinity' should not contain File[/etc/security/limits.d/rabbitmq-server.conf]
rspec ./spec/classes/rabbitmq_spec.rb[1:2:11:2] # rabbitmq on archlinux-4-x86_64 with file_limit => 'infinity' should not contain File[/etc/default/rabbitmq-server]
rspec ./spec/classes/rabbitmq_spec.rb[1:2:11:3] # rabbitmq on archlinux-4-x86_64 with file_limit => 'infinity' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:2:12:1] # rabbitmq on archlinux-4-x86_64 with file_limit => '-1' should not contain File[/etc/security/limits.d/rabbitmq-server.conf]
rspec ./spec/classes/rabbitmq_spec.rb[1:2:12:2] # rabbitmq on archlinux-4-x86_64 with file_limit => '-1' should not contain File[/etc/default/rabbitmq-server]
rspec ./spec/classes/rabbitmq_spec.rb[1:2:12:3] # rabbitmq on archlinux-4-x86_64 with file_limit => '-1' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:2:13:3] # rabbitmq on archlinux-4-x86_64 with file_limit => '1234' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:2:19:1] # rabbitmq on archlinux-4-x86_64 on systems without systemd should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:6:10:1] # rabbitmq on centos-7-x86_64 with file_limit => 'unlimited' should contain File[/etc/security/limits.d/rabbitmq-server.conf] that notifies Class[Rabbitmq::Service]
rspec ./spec/classes/rabbitmq_spec.rb[1:6:10:2] # rabbitmq on centos-7-x86_64 with file_limit => 'unlimited' should not contain File[/etc/default/rabbitmq-server]
rspec ./spec/classes/rabbitmq_spec.rb[1:6:10:3] # rabbitmq on centos-7-x86_64 with file_limit => 'unlimited' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:6:11:1] # rabbitmq on centos-7-x86_64 with file_limit => 'infinity' should contain File[/etc/security/limits.d/rabbitmq-server.conf] that notifies Class[Rabbitmq::Service]
rspec ./spec/classes/rabbitmq_spec.rb[1:6:11:2] # rabbitmq on centos-7-x86_64 with file_limit => 'infinity' should not contain File[/etc/default/rabbitmq-server]
rspec ./spec/classes/rabbitmq_spec.rb[1:6:11:3] # rabbitmq on centos-7-x86_64 with file_limit => 'infinity' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:6:12:1] # rabbitmq on centos-7-x86_64 with file_limit => '-1' should contain File[/etc/security/limits.d/rabbitmq-server.conf] that notifies Class[Rabbitmq::Service]
rspec ./spec/classes/rabbitmq_spec.rb[1:6:12:2] # rabbitmq on centos-7-x86_64 with file_limit => '-1' should not contain File[/etc/default/rabbitmq-server]
rspec ./spec/classes/rabbitmq_spec.rb[1:6:12:3] # rabbitmq on centos-7-x86_64 with file_limit => '-1' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:6:13:3] # rabbitmq on centos-7-x86_64 with file_limit => '1234' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:6:19:1] # rabbitmq on centos-7-x86_64 on systems without systemd should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:7:10:1] # rabbitmq on ubuntu-16.04-x86_64 with file_limit => 'unlimited' should not contain File[/etc/security/limits.d/rabbitmq-server.conf]
rspec ./spec/classes/rabbitmq_spec.rb[1:7:10:2] # rabbitmq on ubuntu-16.04-x86_64 with file_limit => 'unlimited' should contain File[/etc/default/rabbitmq-server] with content =~ /ulimit -n unlimited/
rspec ./spec/classes/rabbitmq_spec.rb[1:7:10:3] # rabbitmq on ubuntu-16.04-x86_64 with file_limit => 'unlimited' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:7:11:1] # rabbitmq on ubuntu-16.04-x86_64 with file_limit => 'infinity' should not contain File[/etc/security/limits.d/rabbitmq-server.conf]
rspec ./spec/classes/rabbitmq_spec.rb[1:7:11:2] # rabbitmq on ubuntu-16.04-x86_64 with file_limit => 'infinity' should contain File[/etc/default/rabbitmq-server] with content =~ /ulimit -n infinity/
rspec ./spec/classes/rabbitmq_spec.rb[1:7:11:3] # rabbitmq on ubuntu-16.04-x86_64 with file_limit => 'infinity' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:7:12:1] # rabbitmq on ubuntu-16.04-x86_64 with file_limit => '-1' should not contain File[/etc/security/limits.d/rabbitmq-server.conf]
rspec ./spec/classes/rabbitmq_spec.rb[1:7:12:2] # rabbitmq on ubuntu-16.04-x86_64 with file_limit => '-1' should contain File[/etc/default/rabbitmq-server] with content =~ /ulimit -n -1/
rspec ./spec/classes/rabbitmq_spec.rb[1:7:12:3] # rabbitmq on ubuntu-16.04-x86_64 with file_limit => '-1' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:7:13:3] # rabbitmq on ubuntu-16.04-x86_64 with file_limit => '1234' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:7:19:1] # rabbitmq on ubuntu-16.04-x86_64 on systems without systemd should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:8:10:1] # rabbitmq on redhat-7-x86_64 with file_limit => 'unlimited' should contain File[/etc/security/limits.d/rabbitmq-server.conf] that notifies Class[Rabbitmq::Service]
rspec ./spec/classes/rabbitmq_spec.rb[1:8:10:2] # rabbitmq on redhat-7-x86_64 with file_limit => 'unlimited' should not contain File[/etc/default/rabbitmq-server]
rspec ./spec/classes/rabbitmq_spec.rb[1:8:10:3] # rabbitmq on redhat-7-x86_64 with file_limit => 'unlimited' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:8:11:1] # rabbitmq on redhat-7-x86_64 with file_limit => 'infinity' should contain File[/etc/security/limits.d/rabbitmq-server.conf] that notifies Class[Rabbitmq::Service]
rspec ./spec/classes/rabbitmq_spec.rb[1:8:11:2] # rabbitmq on redhat-7-x86_64 with file_limit => 'infinity' should not contain File[/etc/default/rabbitmq-server]
rspec ./spec/classes/rabbitmq_spec.rb[1:8:11:3] # rabbitmq on redhat-7-x86_64 with file_limit => 'infinity' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:8:12:1] # rabbitmq on redhat-7-x86_64 with file_limit => '-1' should contain File[/etc/security/limits.d/rabbitmq-server.conf] that notifies Class[Rabbitmq::Service]
rspec ./spec/classes/rabbitmq_spec.rb[1:8:12:2] # rabbitmq on redhat-7-x86_64 with file_limit => '-1' should not contain File[/etc/default/rabbitmq-server]
rspec ./spec/classes/rabbitmq_spec.rb[1:8:12:3] # rabbitmq on redhat-7-x86_64 with file_limit => '-1' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:8:13:3] # rabbitmq on redhat-7-x86_64 with file_limit => '1234' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:8:19:1] # rabbitmq on redhat-7-x86_64 on systems without systemd should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:9:10:1] # rabbitmq on debian-8-x86_64 with file_limit => 'unlimited' should not contain File[/etc/security/limits.d/rabbitmq-server.conf]
rspec ./spec/classes/rabbitmq_spec.rb[1:9:10:2] # rabbitmq on debian-8-x86_64 with file_limit => 'unlimited' should contain File[/etc/default/rabbitmq-server] with content =~ /ulimit -n unlimited/
rspec ./spec/classes/rabbitmq_spec.rb[1:9:10:3] # rabbitmq on debian-8-x86_64 with file_limit => 'unlimited' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:9:11:1] # rabbitmq on debian-8-x86_64 with file_limit => 'infinity' should not contain File[/etc/security/limits.d/rabbitmq-server.conf]
rspec ./spec/classes/rabbitmq_spec.rb[1:9:11:2] # rabbitmq on debian-8-x86_64 with file_limit => 'infinity' should contain File[/etc/default/rabbitmq-server] with content =~ /ulimit -n infinity/
rspec ./spec/classes/rabbitmq_spec.rb[1:9:11:3] # rabbitmq on debian-8-x86_64 with file_limit => 'infinity' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:9:12:1] # rabbitmq on debian-8-x86_64 with file_limit => '-1' should not contain File[/etc/security/limits.d/rabbitmq-server.conf]
rspec ./spec/classes/rabbitmq_spec.rb[1:9:12:2] # rabbitmq on debian-8-x86_64 with file_limit => '-1' should contain File[/etc/default/rabbitmq-server] with content =~ /ulimit -n -1/
rspec ./spec/classes/rabbitmq_spec.rb[1:9:12:3] # rabbitmq on debian-8-x86_64 with file_limit => '-1' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:9:13:3] # rabbitmq on debian-8-x86_64 with file_limit => '1234' should not contain Systemd::Service_limits[rabbitmq.service]
rspec ./spec/classes/rabbitmq_spec.rb[1:9:19:1] # rabbitmq on debian-8-x86_64 on systems without systemd should not contain Systemd::Service_limits[rabbitmq.service]
```

JayH5 commented:
@bastelfreak thanks.. I must've been misreading 😓

wyardley commented:
@JayH5  I don't think this can work with the current systemd module. 
Did you see my comments in:
https://github.com/voxpupuli/puppet-rabbitmq/pull/683#issuecomment-387285180
Since it doesn't seem like that will be fixed soon, probably needs to be implemented w/o the systemd module.

bastelfreak commented:
@wyardley I am happy to fix this in the systemd module. I would like to merge https://github.com/camptocamp/puppet-systemd/pull/81 first and afterwards add support for `-1` and `unlimited`

bastelfreak commented:
I am going to close this in favour of https://github.com/voxpupuli/puppet-rabbitmq/pull/715

antondollmaier commented on manifests/config.pp:
```diff
@@ -255,28 +213,14 @@
+  }
+
+  if $facts['systemd'] { # systemd fact provided by systemd module
+    systemd::service_limits { 'rabbitmq.service':
```
Shouldn't this use $::rabbitmq::service_name?

On Debian Stretch, the Service is called "rabbitmq-server.service" - but with "rabbitmq.service", the override does not get applied.
